File: scripts/prediction-sim/lib/retroactive-apply-plan.mjs
// Pure-logic JS port of keyboard/Sources/Prediction/Autocorrect/RetroactiveApplyPlan.swift.
// Kept in sync per AGENTS.md -> Test policy.

/**
 * Builds a cursor-relative plan for replacing a typed word with a correction.
 *
 * Mirrors RetroactiveApplyPlan.plan(typedWord:correction:offsetFromCursorEnd:)
 * in Swift. Expressed as textDocumentProxy operations applied in order:
 *   1. adjustTextPosition(byCharacterOffset: backMove)  — move the cursor back
 *      to the end of the typed word.
 *   2. deleteBackward() × deleteCount                    — remove the typed word.
 *   3. insertText(insert)                                — place the correction.
 *   4. adjustTextPosition(byCharacterOffset: forwardMove) — return the cursor
 *      to its original relative spot.
 *
 * INVARIANT: forwardMove === offsetFromCursorEnd REGARDLESS of any length
 * difference between typedWord and correction — the delete + insert happens
 * entirely between the cursor's back-move target and the cursor's original
 * position, so the cursor's RELATIVE distance from the corrected word's end is
 * preserved.
 *
 * @param {string} typedWord - The word the user actually typed.
 * @param {string} correction - The correction text to insert.
 * @param {number} offsetFromCursorEnd - Char distance from the cursor to the
 *   typed word's body end.
 * @returns {{backMove: number, deleteCount: number, insert: string, forwardMove: number}}
 */
export function plan(typedWord, correction, offsetFromCursorEnd) {
  return {
    backMove: -offsetFromCursorEnd,
    deleteCount: typedWord.length,
    insert: correction,
    forwardMove: offsetFromCursorEnd,
  };
}

File: keyboard/Sources/Prediction/Autocorrect/RetroactiveApplyPlan.swift
import Foundation

/// Pure offset arithmetic for replacing a typed word with a correction at a
/// cursor-relative location. No UIKit import.
///
/// The plan is expressed as `textDocumentProxy` operations applied in order:
///   1. `adjustTextPosition(byCharacterOffset: backMove)` — move the cursor
///      back to the end of the typed word.
///   2. `deleteBackward()` × `deleteCount` — remove the typed word.
///   3. `insertText(insert)` — place the correction.
///   4. `adjustTextPosition(byCharacterOffset: forwardMove)` — return the
///      cursor to its original relative spot.
enum RetroactiveApplyPlan {

    /// A cursor-relative replacement plan.
    struct Plan {
        /// Negative offset: move the cursor back to the end of the typed word.
        let backMove: Int
        /// Number of characters to delete (always `typedWord.count`).
        let deleteCount: Int
        /// The correction text to insert.
        let insert: String
        /// Positive offset: move the cursor forward again after inserting.
        let forwardMove: Int
    }

    /// Builds a plan for replacing `typedWord` with `correction` when the typed
    /// word's body ends `offsetFromCursorEnd` characters before the cursor.
    ///
    /// INVARIANT: `forwardMove == offsetFromCursorEnd` REGARDLESS of any length
    /// difference between `typedWord` and `correction`. The delete + insert
    /// happens entirely between the cursor's back-move target and the cursor's
    /// original position, so the cursor's RELATIVE distance from the corrected
    /// word's end is preserved — the cursor ends up at the same relative spot
    /// whether the correction is shorter, longer, or the same length as the
    /// typed word. (In absolute terms the cursor shifts by the length delta,
    /// which is exactly what keeps it aligned with the corrected word.)
    static func plan(typedWord: String, correction: String, offsetFromCursorEnd: Int) -> Plan {
        Plan(
            backMove: -offsetFromCursorEnd,
            deleteCount: typedWord.count,
            insert: correction,
            forwardMove: offsetFromCursorEnd
        )
    }
}
